File: algorithms/algorithms-cplusplus/leetcode/Question_0152_Maximum_Product_Subarray.cc
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
 public:
  int maxProduct(vector<int>& nums) {
    if (nums.empty()) {
      return 0;
    }
    int pre_min = nums[0], pre_max = nums[0], res = nums[0];
    int cur_min, cur_max;
    for (int i = 1; i < nums.size(); i++) {
      int n = nums[i];
      cur_min = min(n, min(pre_min * n, pre_max * n));
      cur_max = max(n, max(pre_min * n, pre_max * n));
      pre_min = cur_min;
      pre_max = cur_max;
      res = max(res, cur_max);
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0152_Maximum_Product_Subarray.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0152_Maximum_Product_Subarray {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int preMax = nums[0], preMin = nums[0], curMax = 0, curMin = 0, res = nums[0];
        for (int i = 1; i < nums.length; i++) {
            curMax = Math.max(nums[i], Math.max(preMax * nums[i], preMin * nums[i]));
            curMin = Math.min(nums[i], Math.min(preMax * nums[i], preMin * nums[i]));
            res = Math.max(res, curMax);
            preMax = curMax;
            preMin = curMin;
        }
        return res;
    }
}
